src/replit_river/server_session.py

```diff
@@ -0,0 +1,247 @@
+                    if self._retry_connection_callback:
+                        self._task_manager.create_task(
+                            self._retry_connection_callback()
+                        )
+
+                    await self._begin_close_session_countdown()
+                    logger.debug("ConnectionClosed while serving", exc_info=True)
+                except FailedSendingMessageException:
+                    # Expected error if the connection is closed.
+                    logger.debug(
+                        "FailedSendingMessageException while serving", exc_info=True
+                    )
+                except Exception:
+                    logger.exception("caught exception at message iterator")
+        except ExceptionGroup as eg:
+            _, unhandled = eg.split(lambda e: isinstance(e, ConnectionClosed))
+            if unhandled:
+                raise ExceptionGroup(
+                    "Unhandled exceptions on River server", unhandled.exceptions
+                )
+
+    async def _handle_messages_from_ws(self, tg: asyncio.TaskGroup) -> None:
+        logger.debug(
+            "%s start handling messages from ws %s",
+            "server",
+            self._ws_wrapper.id,
+        )
+        try:
+            ws_wrapper = self._ws_wrapper
+            async for message in ws_wrapper.ws:
+                try:
+                    if not await ws_wrapper.is_open():
+                        # We should not process messages if the websocket is closed.
+                        break
+                    msg = parse_transport_msg(message, self._transport_options)
+
+                    logger.debug(f"{self._transport_id} got a message %r", msg)
+
+                    # Update bookkeeping
+                    await self._seq_manager.check_seq_and_update(msg)
+                    await self._buffer.remove_old_messages(
+                        self._seq_manager.receiver_ack,
+                    )
+                    self._reset_session_close_countdown()
+
+                    if msg.controlFlags & ACK_BIT != 0:
+                        continue
+                    async with self._stream_lock:
+                        stream = self._streams.get(msg.streamId, None)
+                    if msg.controlFlags & STREAM_OPEN_BIT == 0:
+                        if not stream:
+                            logger.warning("no stream for %s", msg.streamId)
+                            raise IgnoreMessageException(
+                                "no stream for message, ignoring"
+                            )
+                        await add_msg_to_stream(msg, stream)
+                    else:
+                        # TODO(dstewart) This looks like it opens a new call to handler
+                        #                on ever ws message, instead of demuxing and
+                        #                routing.
+                        _stream = await self._open_stream_and_call_handler(msg, tg)
```

Comment:
:eyes:

Reply:
Yeah Connor found that as well -- the comment is stale from when I was first getting up to speed on river